src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java

```diff
@@ -0,0 +1,247 @@
+        try {
+            engine.deliver(ctx,
+                           name,
+
+                           objectId,
+                           fileExtension,
+                           status -> handleHttpError(ctx, objectId, fileExtension, null, status));
+        } catch (IOException e) {
+            throw Exceptions.handle()
+                            .error(e)
+                            .to(StorageUtils.LOG)
+                            .withSystemErrorMessage(
+                                    "Storage Layer 1: An error occurred when delivering %s (%s) for %s: %s (%s)",
+                                    objectId,
+                                    name,
+                                    ctx.getRequestedURI())
+                            .handle();
+        }
+    }
+
+    /**
+     * Delivers the requested object to the given request as response.
+     * <p>
+     * If replication is active and delivery from the primary storage fails a delivery from the backup space is
+     * attempted automatically (for 5XX HTTP errors).
+     *
+     * @param ctx      the request to provide a response for
+     * @param objectId the id of the object to deliver
+     * @param filename the file name to specify
+     */
+    public void deliverAsDownload(WebContext ctx, String objectId, String filename) {
+        try {
+            engine.deliverAsDownload(ctx,
+                                     name,
+
+                                     objectId,
+                                     filename,
+                                     status -> handleHttpError(ctx, objectId, null, filename, status));
+        } catch (IOException e) {
+            try {
+                ctx.respondWith().error(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+            } catch (Exception ex) {
+                Exceptions.ignore(ex);
+            }
+            throw Exceptions.handle()
+                            .error(e)
+                            .to(StorageUtils.LOG)
+                            .withSystemErrorMessage(
+                                    "Storage Layer 1: An error occurred when delivering %s (%s) for %s: %s (%s)",
+                                    objectId,
+                                    name,
+                                    ctx.getRequestedURI())
+                            .handle();
+        }
+    }
+
+    private void handleHttpError(WebContext ctx,
+                                 String objectId,
+                                 @Nullable String fileExtension,
+                                 @Nullable String filename,
+                                 int status) {
+        PhysicalStorageSpace replicationSpace = replicationManager.getReplicationSpace(name).orElse(null);
+        if (replicationSpace == null) {
+            ctx.respondWith().error(HttpResponseStatus.valueOf(status));
+            //TODO bump stats + record event
```

Comment:
Link ticket or will this be done in one of the followup PRs?

Reply:
followup